src/test_fixture/transport/channeled_transport.rs

```diff
@@ -0,0 +1,575 @@
+            }),
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::ff::FieldType;
+    use crate::helpers::query::QueryType;
+    use crate::helpers::HelperIdentity;
+    use crate::protocol::Step;
+    use futures_util::stream::poll_immediate;
+    use tokio::sync::mpsc::channel;
+    use tokio::sync::oneshot;
+
+    const STEP: &str = "in-memory-transport";
+
+    fn stub_callbacks() -> TransportCallbacks<impl ReceiveQueryCallback> {
+        TransportCallbacks {
+            receive_query: move |_| Box::pin(async { unimplemented!() }),
+        }
+    }
+
+    fn one_transport_setup<I: Into<i32>, C: ReceiveQueryCallback + 'static>(
+        id: I,
+        cb: TransportCallbacks<C>,
+    ) -> Setup<C> {
+        let id = HelperIdentity::from(id.into());
+
+        Setup::new(id, cb)
+    }
+
+    fn one_transport<I: Into<i32>, C: ReceiveQueryCallback + 'static>(
+        id: I,
+        cb: TransportCallbacks<C>,
+    ) -> (ConnectionTx, InMemoryChannelledTransport) {
+        one_transport_setup(id, cb).finish()
+    }
+
+    #[tokio::test]
+    async fn callback_is_called() {
+        let (signal_tx, signal_rx) = oneshot::channel();
+        let signal_tx = Arc::new(Mutex::new(Some(signal_tx)));
+        let (tx, _transport) = one_transport(
+            1,
+            TransportCallbacks {
+                receive_query: move |query_config| {
+                    Box::pin({
+                        let signal_tx = Arc::clone(&signal_tx);
+                        async move {
+                            // this works because callback is only called once
+                            signal_tx
+                                .lock()
+                                .unwrap()
+                                .take()
+                                .unwrap()
+                                .send(query_config)
+                                .unwrap();
+                            Ok(QueryId)
+                        }
+                    })
+                },
+            },
+        );
+        let expected = QueryConfig {
+            field_type: FieldType::Fp32BitPrime,
+            query_type: QueryType::TestMultiply,
+        };
+        tx.send((
+            InMemoryPacket::receive_query(expected),
+            InMemoryStream::empty(),
+        ))
+        .await
+        .unwrap();
+
+        assert_eq!(expected, signal_rx.await.unwrap());
+    }
+
+    #[tokio::test]
+    async fn receive_not_ready() {
+        let (tx, transport) = one_transport(1, stub_callbacks());
+        let from = HelperIdentity::from(2);
+        let expected = vec![vec![1], vec![2]];
+
+        let mut stream = transport
+            .receive(from, (QueryId, Step::from(STEP)))
+            .unwrap();
+
+        // make sure it is not ready as it hasn't received the records stream yet.
+        assert!(matches!(
+            poll_immediate(&mut stream).next().await,
+            Some(Poll::Pending)
+        ));
```

Comment:
Would this be better?

```suggestion
assert!(stream.next().now_or_never().is_none());
```

BTW, the [documentation](https://docs.rs/futures/latest/futures/stream/struct.PollImmediate.html) for `PollImmediate` is really confusing.  It's not immediately obvious that when it wraps `S`, the item it returns is `Option<Poll<S::Item>>`.

Reply:
yep that is better except in one case - it does not distinguish between closed and not ready states. I need it in one test, so `poll_immediate` is still better there imo

Reply:
funny, but I just realized I can't use `now_or_never` because of `old_waker.will_wake(waker)` check. `now_or_never` constructs a new, no-op, waker while `poll_immediate` does not need tht